app/models/ability.rb

```diff
@@ -138,4 +122,49 @@ def initialize(user)
     can :index, GogovanTransport
     can :index, CrossroadsTransport
   end
+
+  def taxonomies
+    can :register, :device
+    can [:index, :show], DonorCondition
+    can [:index, :show], SubpackageType
+    can [:index, :show], RejectionReason
+    can [:index, :show], Permission
+
+    # TODO
+    can [:create, :show], Address
+    can [:create, :destroy], Contact
+
+    # Schedule
+    # TODO - only for offers owned by user
+    can :create, Schedule
+
+    # GogovanOrder
+    # TODO - only for offers owned by user
+    can [:calculate_price, :confirm_order, :destroy], GogovanOrder
+  end
+
+  def user_abilities
+    can [:show, :update], User, id: @user_id
+    can [:index, :show, :update], User if staff?
+    can :current_user_profile, User
+  end
+
+  def version_abilities
+    can :index, Version, related_type: "Offer", related_id: @user_offer_ids
+    can :index, Version, item_type: "Offer", item_id: @user_offer_ids
+    can :index, Version if staff?
+  end
+
+  private
+
+  def staff?
+    @reviewer || @supervisor
+  end
+
+  def action_ability(actions, class_name, records)
+    can actions, class_name, records do |record|
+      record.item.offer.created_by_id == @user_id
+    end
+  end
 end
+
```

Comment:
@swatijadhav Overall, I like what you've done here as it separates various sections into their own function. I think `csui` and `action_ability` makes things a bit harder to read by hiding implementation. If you could revert just those bits, I think this will be ok. When doing a security audit, I really want to see the logic all together so it's clear exactly what each permission covers.